core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts

```diff
@@ -0,0 +1,419 @@
+        .onDiagnosticTestSessionCompleted.emit(recommendedTopicIds);
+    });
+  }
+
+  getRecommendedTopicIds(): string[] {
+    let recommendedTopicIds: string[] = [];
+    let failedTopicIds: string[] = this.getFailedTopicIds();
+    let rootTopicIds: string[] = this.getRootTopicIds();
+    let failedRootTopicIds: string[] = rootTopicIds.filter(
+      topicId => failedTopicIds.indexOf(topicId) !== -1
+    );
+
+    if (failedRootTopicIds.length >= 2) {
+      recommendedTopicIds.push(failedRootTopicIds[0]);
+      recommendedTopicIds.push(failedRootTopicIds[1]);
+    } else if (failedRootTopicIds.length === 1) {
+      recommendedTopicIds.push(failedRootTopicIds[0]);
+    } else {
+      let sortedTopicIds = this.getSortedTopicIds();
+      for (let topicId of sortedTopicIds) {
+        if (failedTopicIds.indexOf(topicId) !== -1) {
+          recommendedTopicIds.push(topicId);
+          break;
+        }
+      }
+    }
+    return recommendedTopicIds;
+  }
+
+  getRootTopicIds(): string[] {
+    let topicIdToPrerequisiteTopicId = (
+      this._initialCopyOfTopicTrackerModel.getTopicIdToPrerequisiteTopicIds());
+    let rootTopicIds: string[] = [];
+
+    for (let topicId in topicIdToPrerequisiteTopicId) {
+      if (topicIdToPrerequisiteTopicId[topicId].length === 0) {
+        rootTopicIds.push(topicId);
+      }
+    }
+    return rootTopicIds;
+  }
+
+  getSortedTopicIds(): string[] {
+    let visitedTopicIds: string[] = [];
+    let topicIdToPrerequisiteTopicId = (
+      this._initialCopyOfTopicTrackerModel.getTopicIdToPrerequisiteTopicIds());
+
+    for (let currentTopicId in topicIdToPrerequisiteTopicId) {
+      if (visitedTopicIds.indexOf(currentTopicId) !== -1) {
+        continue;
+      }
+
+      let tempStack = [];
+      tempStack.push(currentTopicId);
+
+      while (tempStack.length > 0) {
+        // '.shift()' here can return an undefined value, but we're already
+        // checking for length > 0, so this is safe.
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        let topicId: string = tempStack.shift()!;
+        let prerequisites: string[] = topicIdToPrerequisiteTopicId[topicId];
+        let nonVisitedPrerequisites: string[] = prerequisites.filter(
+          (prerequisiteTopicId: string) => {
+            return visitedTopicIds.indexOf(prerequisiteTopicId) === -1;
+          }
+        );
+
+        if (nonVisitedPrerequisites.length > 0) {
+          tempStack.unshift(topicId);
+          tempStack = nonVisitedPrerequisites.concat(tempStack);
+        } else {
+          visitedTopicIds.push(topicId);
+        }
+      }
+    }
+    return visitedTopicIds;
+  }
+
+  computeProgressPercentage(): number {
+    let numberOfAttemptedQuestionsInCurrentTopic = (
+      this._diagnosticTestCurrentTopicStatusModel.numberOfAttemptedQuestions);
+    let initialTopicIdsList = (
+      this._initialCopyOfTopicTrackerModel.getEligibleTopicIds());
```

Comment:
I thought we were trying to move away from the use of the term "eligible"?

Reply:
Removed